### PR TITLE
Add support for CapacityRebalance

### LINKED ISSUE #16033: Support EC2 rebalancing
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS introduced EC2 rebalancing for ASG, EC2 Fleet and Spot Fleet.

### New or Affected Resource(s)

* aws_autoscaling_group
* aws_spot_fleet

### References

See: https://aws.amazon.com/about-aws/whats-new/2020/11/introducing-ec2-instance-rebalance-recommendation-for-ec2-spot-instances/

Flag: https://docs.aws.amazon.com/autoscaling/ec2/userguide/capacity-rebalance.html

### PR DESCRIPTION
**Pull request by AdamTylerLynch**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #16033

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```
resource/aws_autoscaling_group: Add support for CapacityRebalance
resource/aws_ec2_fleet: Add support for Maintenance Strategies of CapacityRebalance
resource/aws_spot_fleet_request: Add support for Spot Maintenance Strategies of CapacityRebalance
docs/autoscaling_group:  Added documentation and usage sample for CapacityRebalance
docs/ec2_fleet: Added documentation for Maintenance Strategies of Capacity Rebalance
docs/spot_fleet_request: Add documentation for Spot Maintenance Strategies of CapacityRebalance

```

Output from acceptance testing:

```
make testacc TESTARGS='-run=TestAccAWSAutoScalingGroup_MixedInstancesPolicy_CapacityRebalance'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSAutoScalingGroup_MixedInstancesPolicy_CapacityRebalance -timeout 120m
=== RUN   TestAccAWSAutoScalingGroup_MixedInstancesPolicy_CapacityRebalance
=== PAUSE TestAccAWSAutoScalingGroup_MixedInstancesPolicy_CapacityRebalance
=== CONT  TestAccAWSAutoScalingGroup_MixedInstancesPolicy_CapacityRebalance
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_CapacityRebalance (37.77s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	40.555s



make testacc TESTARGS='-run=TestAccAWSEc2Fleet_SpotOptions_CapacityRebalance'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSEc2Fleet_SpotOptions_CapacityRebalance -timeout 120m
=== RUN   TestAccAWSEc2Fleet_SpotOptions_CapacityRebalance
=== PAUSE TestAccAWSEc2Fleet_SpotOptions_CapacityRebalance
=== CONT  TestAccAWSEc2Fleet_SpotOptions_CapacityRebalance
--- PASS: TestAccAWSEc2Fleet_SpotOptions_CapacityRebalance (39.20s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	41.899s



make testacc TESTARGS='-run=TestAccAWSSpotFleetRequest_CapacityRebalance'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSpotFleetRequest_CapacityRebalance -timeout 120m
=== RUN   TestAccAWSSpotFleetRequest_CapacityRebalance
=== PAUSE TestAccAWSSpotFleetRequest_CapacityRebalance
=== CONT  TestAccAWSSpotFleetRequest_CapacityRebalance
--- PASS: TestAccAWSSpotFleetRequest_CapacityRebalance (124.75s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	127.486s

```

### CODE REVIEW

**Comment by AdamTylerLynch** [comment]
Requires AWS SDK v1.35.21:

#16021

**Comment by AdamTylerLynch** [comment]
```
aws_autoscaling_group
resource "aws_autoscaling_group" "example" {
  capacity_rebalance = true
}

aws_ec2_fleet
resource "aws_ec2_fleet" "example" {
  spot_options {
    maintenance_strategies {
      capacity_rebalance {
        replacement_strategy = "launch"
      }
    }
  }
}

aws_spot_fleet_request
resource "aws_spot_fleet_request" "example" {
  spot_maintenance_strategies {
    capacity_rebalance {
      replacement_strategy = "launch"
    }
  }
}
```

File: aws/resource_aws_ec2_fleet_test.go
Comment (ewbankkit) [inline aws/resource_aws_ec2_fleet_test.go:1805]: Formatting

```suggestion
  spot_options {
    allocation_strategy = %q
    maintenance_strategies {
      capacity_rebalance {
        replacement_strategy = "launch"
      }
    }	
  }
```

File: aws/resource_aws_spot_fleet_request_test.go
Comment (ewbankkit) [inline aws/resource_aws_spot_fleet_request_test.go:2472]: Formatting

```suggestion
  spot_maintenance_strategies {
    capacity_rebalance {
      replacement_strategy = "launch"
    }
  }	
```

File: website/docs/r/autoscaling_group.html.markdown
Comment (ewbankkit) [inline website/docs/r/autoscaling_group.html.markdown:156]: Formatting

```suggestion
    instance_distribution {
      on_demand_base_capacity                  = 0
      on_demand_percentage_above_base_capacity = 25
      spot_allocation_strategy                 = "capacity-optimized"
    }
```

File: aws/resource_aws_autoscaling_group.go
Comment (ewbankkit) [inline aws/resource_aws_autoscaling_group.go:210]: No need for `Default: false` (that's the default).

Comment (AdamTylerLynch) [reply aws/resource_aws_autoscaling_group.go]: Resolved in cf0e240

File: aws/resource_aws_ec2_fleet_test.go
Comment (ewbankkit) [inline aws/resource_aws_ec2_fleet_test.go:1804]: Extra whitespace at EOL.

```suggestion
    }
```

File: aws/resource_aws_spot_fleet_request_test.go
Comment (ewbankkit) [inline aws/resource_aws_spot_fleet_request_test.go:2472]: Extra whitespace at EOL.

```suggestion
  }
```

File: website/docs/r/autoscaling_group.html.markdown
Comment (ewbankkit) [inline website/docs/r/autoscaling_group.html.markdown:149]: Alignment.

```suggestion
  capacity_rebalance   = true
  desired_capacity     = 12
  max_size             = 15
  min_size             = 12
  vpc_zone_identifier  = [aws_subnet.example1.id, aws_subnet.example2.id]
```
```suggestion
  capacity_rebalance = true
  desired_capacity   = 12
  max_size           = 15
  min_size           = 12
  vpc_zone_identifier  = [aws_subnet.example1.id, aws_subnet.example2.id]
```

File: aws/resource_aws_ec2_fleet_test.go
Comment (AdamTylerLynch) [reply aws/resource_aws_ec2_fleet_test.go:1804]: Removed unneeded spacing.

File: website/docs/r/autoscaling_group.html.markdown
Comment (ewbankkit) [inline website/docs/r/autoscaling_group.html.markdown:149]: Really fix alignment 😄.

```suggestion
  capacity_rebalance  = true
  desired_capacity    = 12
  max_size            = 15
  min_size            = 12
  vpc_zone_identifier = [aws_subnet.example1.id, aws_subnet.example2.id]
```

**Comment by ewbankkit** [comment]
Verified acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSAutoScalingGroup_MixedInstancesPolicy_CapacityRebalance\|TestAccAWSAutoScalingGroup_basic\|TestAccAWSEc2Fleet_SpotOptions_CapacityRebalance\|TestAccAWSEc2Fleet_basic\|TestAccAWSSpotFleetRequest_CapacityRebalance\|TestAccAWSSpotFleetRequest_basic' ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 3 -run=TestAccAWSAutoScalingGroup_MixedInstancesPolicy_CapacityRebalance\|TestAccAWSAutoScalingGroup_basic\|TestAccAWSEc2Fleet_SpotOptions_CapacityRebalance\|TestAccAWSEc2Fleet_basic\|TestAccAWSSpotFleetRequest_CapacityRebalance\|TestAccAWSSpotFleetRequest_basic -timeout 120m
=== RUN   TestAccAWSAutoScalingGroup_basic
=== PAUSE TestAccAWSAutoScalingGroup_basic
=== RUN   TestAccAWSAutoScalingGroup_MixedInstancesPolicy_CapacityRebalance
=== PAUSE TestAccAWSAutoScalingGroup_MixedInstancesPolicy_CapacityRebalance
=== RUN   TestAccAWSEc2Fleet_basic
=== PAUSE TestAccAWSEc2Fleet_basic
=== RUN   TestAccAWSEc2Fleet_SpotOptions_CapacityRebalance
=== PAUSE TestAccAWSEc2Fleet_SpotOptions_CapacityRebalance
=== RUN   TestAccAWSSpotFleetRequest_basic
=== PAUSE TestAccAWSSpotFleetRequest_basic
=== RUN   TestAccAWSSpotFleetRequest_CapacityRebalance
=== PAUSE TestAccAWSSpotFleetRequest_CapacityRebalance
=== CONT  TestAccAWSAutoScalingGroup_basic
=== CONT  TestAccAWSSpotFleetRequest_CapacityRebalance
=== CONT  TestAccAWSSpotFleetRequest_basic
--- PASS: TestAccAWSSpotFleetRequest_CapacityRebalance (124.04s)
=== CONT  TestAccAWSEc2Fleet_basic
--- PASS: TestAccAWSSpotFleetRequest_basic (124.62s)
=== CONT  TestAccAWSEc2Fleet_SpotOptions_CapacityRebalance
=== CONT  TestAccAWSAutoScalingGroup_MixedInstancesPolicy_CapacityRebalance
--- PASS: TestAccAWSEc2Fleet_SpotOptions_CapacityRebalance (39.14s)
--- PASS: TestAccAWSEc2Fleet_basic (39.86s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_CapacityRebalance (40.09s)
--- PASS: TestAccAWSAutoScalingGroup_basic (207.22s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	207.295s
```

File: aws/resource_aws_autoscaling_group.go
Comment (bill-rich) [inline aws/resource_aws_autoscaling_group.go:925]: This should be set even if `ok==false` otherwise going from `true` to removing the attribute will leave it `true`.

File: aws/resource_aws_ec2_fleet.go
Comment (bill-rich) [inline aws/resource_aws_ec2_fleet.go:210]: Want to confirm here that the idea is that if `maintenance_strategies` or `capacity_rebalance` is removed from the configuration, that should be ignored (not removed through the API)?

**Review by bill-rich — CHANGES_REQUESTED** [review]

File: aws/resource_aws_ec2_fleet.go
Comment (AdamTylerLynch) [reply aws/resource_aws_ec2_fleet.go:210]: I would defer to the maintainers to determine the behavior that would be considered expected. If configurations are normally removed, is the user expectation to reset them?

File: aws/resource_aws_autoscaling_group.go
Comment (AdamTylerLynch) [reply aws/resource_aws_autoscaling_group.go:925]: Understood. Will submit change shortly.

File: aws/resource_aws_ec2_fleet.go
Comment (AdamTylerLynch) [reply aws/resource_aws_ec2_fleet.go:210]: I copied the logic that I saw used for spot_options and on_demand_options. I am open to suggestions on the best way to handle this.

Comment (AdamTylerLynch) [reply aws/resource_aws_ec2_fleet.go:210]: @bill-rich would you be willing to also evaluate the spot_maintenance_strategies definition in the resource_aws_spot_fleet_request.go ?

File: aws/resource_aws_autoscaling_group.go
Comment (AdamTylerLynch) [reply aws/resource_aws_autoscaling_group.go:927]: Addressed in commit bb483e8

File: aws/resource_aws_ec2_fleet.go
Comment (ewbankkit) [reply aws/resource_aws_ec2_fleet.go:210]: @AdamTylerLynch Some context on this type of diff suppression (see #7180):

```go
// suppressMissingOptionalConfigurationBlock handles configuration block attributes in the following scenario:
//  * The resource schema includes an optional configuration block with defaults
//  * The API response includes those defaults to refresh into the Terraform state
//  * The operator's configuration omits the optional configuration block
```

**Review by bill-rich — APPROVED** [review]
Thanks for addressing the comments. Looks good!

**Pull request merged**
